src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs

```diff
@@ -95,9 +95,43 @@ internal static partial class CertificateValidationPal
             return result;
         }
 
-        // This is only called when we selected local client certificate.
-        // Currently this is only when OpenSSL needs it because peer asked.
-        internal static bool IsLocalCertificateUsed(SafeFreeCredentials? _1, SafeDeleteContext? _2) => true;
+        internal static bool IsLocalCertificateUsed(SafeFreeCredentials? _1, SafeDeleteContext? ctx)
+        {
+            if (ctx is not SafeSslHandle ssl)
+            {
+                return false;
+            }
+
+            if (!Interop.Ssl.SslSessionReused(ssl))
+            {
+                // Fresh session, we set the certificate on the SSL object only
+                // if the peer explicitly requested it.
+                return Interop.Ssl.SslGetCertificate(ssl) != IntPtr.Zero;
+            }
+
+            // resumed session, we keep the information about cert being used in the SSL_SESSION
+            // object's ex_data
+            bool addref = false;
+            try
+            {
+                // make sure the ssl is not freed while we accessing its SSL_SESSION
+                // this makes sure the `session` pointer is valid during this call
+                // despite not being a SafeHandle.
+                ssl.DangerousAddRef(ref addref);
```

Comment:
We don't have SafeHandle type (yet) for SSL_SESSION and here would be the only usage that would need it. Unless you think we should go the extra mile to be sure, I would like to keep it this way to avoid the extra allocations and P/Invokes for the SafeHandle.

Reply:
I think I just got confused by the comment.  It's doing the AddRef/Release pattern that a P/Invoke in terms of a SafeHandle already does, and the comment sounded like "we need to do this addref/release because we're going to pass `ssl.DangerousGetHandle()`"; but instead it's that you want the SSL* to stay up-refd while you get, use, and discard the SSL_SESSION*.  Which the comment does say, upon a re-read.